src/zenml/zen_server/routers/steps_endpoints.py

```diff
@@ -200,6 +203,59 @@ def update_step(
     return dehydrate_response_model(updated_step)
 
 
+@router.put(
+    "/{step_run_id}" + HEARTBEAT,
+    responses={401: error_response, 404: error_response, 422: error_response},
+)
+@async_fastapi_endpoint_wrapper(deduplicate=True)
+def update_heartbeat(
+    step_run_id: UUID,
+    auth_context: AuthContext = Security(authorize),
+) -> StepHeartbeatResponse:
+    """Updates a step.
+
+    Args:
+        step_run_id: ID of the step.
+        auth_context: Authorization/Authentication context.
+
+    Returns:
+        The step heartbeat response (id, status, last_heartbeat).
+    """
+    step = zen_store().get_run_step(step_run_id, hydrate=False)
+
+    if step.status.is_finished:
+        raise HTTPException(
+            status_code=422,
+            detail=f"Step {step.id} is finished - can not update heartbeat.",
+        )
+
+    def validate_token_access(ctx: AuthContext, step_: StepRunResponse):
+        if ctx.access_token.pipeline_run_id:
+            if step_.pipeline_run_id != ctx.access_token.pipeline_run_id:
+                raise AuthorizationException(
+                    f"Authentication token provided is invalid for step: {step_.id}"
+                )
+        elif ctx.access_token.schedule_id:
+            pipeline_run = zen_store().get_run(
+                step_.pipeline_run_id, hydrate=False
+            )
+            schedule = zen_store().get_schedule(
+                ctx.access_token.schedule_id, hydrate=True
+            )
+
+            if pipeline_run.pipeline.id != schedule.pipeline_id:
```

Comment:
Maybe instead of this we can check `pipeline_run.schedule.id == ctx.access_token.schedule_id`, to make sure the run is actually triggered by the schedule that the token is scoped to?